src/z_data/z_expr.rs

```diff
@@ -0,0 +1,194 @@
+            ZExpr::Num(f) => ZPtr(ExprTag::Num, *f),
+            ZExpr::StrNil => ZPtr(ExprTag::Str, F::ZERO),
+            ZExpr::StrCons(x, y) => ZPtr(
+                ExprTag::Str,
+                cache.hash4(&[x.0.to_field(), x.1, y.0.to_field(), y.1]),
+            ),
+            ZExpr::Thunk(x, y) => ZPtr(
+                ExprTag::Thunk,
+                cache.hash4(&[x.0.to_field(), x.1, y.0.to_field(), y.1]),
+            ),
+            ZExpr::Char(f) => ZPtr(ExprTag::Char, F::from_char(*f)),
+            ZExpr::UInt(x) => match x {
+                UInt::U64(x) => ZPtr(ExprTag::U64, F::from_u64(*x)),
+            },
+        }
+    }
+
+    pub fn from_ptr(store: &Store<F>, ptr: &Ptr<F>) -> Option<Self> {
+        match ptr.tag {
+            ExprTag::Nil => Some(ZExpr::Nil),
+            ExprTag::Cons => store.fetch_cons(ptr).and_then(|(car, cdr)| {
+                if let (Some(car), Some(cdr)) = (store.hash_expr(car), store.hash_expr(cdr)) {
+                    Some(ZExpr::Cons(car, cdr))
+                } else {
+                    None
+                }
+            }),
+            ExprTag::Comm => store.fetch_comm(ptr).and_then(|(secret, payload)| {
+                store
+                    .hash_expr(payload)
+                    .map(|payload| ZExpr::Comm(secret.0, payload))
+            }),
+            ExprTag::Sym => store.fetch_symcons(ptr).and_then(|(tag, val)| {
+                if let (Some(tag), Some(val)) = (store.hash_expr(&tag), store.hash_expr(&val)) {
+                    Some(ZExpr::SymCons(tag, val))
+                } else {
+                    None
+                }
+            }),
+            ExprTag::Key => store.hash_expr(ptr).map(ZExpr::Key),
+            ExprTag::Fun => store.fetch_fun(ptr).and_then(|(arg, body, closed_env)| {
+                if let (Some(arg), Some(body), Some(closed_env)) = (
+                    store.hash_expr(arg),
+                    store.hash_expr(body),
+                    store.hash_expr(closed_env),
+                ) {
+                    Some(ZExpr::Fun {
+                        arg,
+                        body,
+                        closed_env,
+                    })
+                } else {
+                    None
+                }
+            }),
+            ExprTag::Num => store.fetch_num(ptr).map(|num| match num {
+                Num::U64(x) => ZExpr::Num((*x).into()),
+                Num::Scalar(x) => ZExpr::Num(*x),
+            }),
+            ExprTag::Str => store.fetch_strcons(ptr).and_then(|(tag, val)| {
+                if let (Some(tag), Some(val)) = (store.hash_expr(&tag), store.hash_expr(&val)) {
+                    Some(ZExpr::StrCons(tag, val))
+                } else {
+                    None
+                }
+            }),
+            ExprTag::Char => store.fetch_char(ptr).map(ZExpr::Char),
+            ExprTag::U64 => store.fetch_uint(ptr).map(ZExpr::UInt),
+            ExprTag::Thunk => unimplemented!(),
```

Comment:
Having any missing parts in de/serialization is going to hurt because inevitably we'll need it at some point and then get frustrated that it's missing. Let's just be comprehensive once and for all.